Code/Source/solver/integrator.cpp

```diff
@@ -0,0 +1,361 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+ * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "integrator.h"
+#include "all_fun.h"
+#include "bf.h"
+#include "contact.h"
+#include "eq_assem.h"
+#include "fs.h"
+#include "ls.h"
+#include "output.h"
+#include "pic.h"
+#include "ris.h"
+#include "set_bc.h"
+#include "ustruct.h"
+
+#include <algorithm>
+#include <iostream>
+#include <set>
+
+using namespace consts;
+
+//------------------------
+// Integrator Constructor
+//------------------------
+Integrator::Integrator(Simulation* simulation)
+  : simulation_(simulation), inner_count_(0)
+{
+  initialize_arrays();
+}
+
+//------------------------
+// Integrator Destructor
+//------------------------
+Integrator::~Integrator() {
+  // Arrays will be automatically cleaned up
+}
+
+//------------------------
+// initialize_arrays
+//------------------------
+void Integrator::initialize_arrays() {
+  auto& com_mod = simulation_->com_mod;
+  int tDof = com_mod.tDof;
+  int tnNo = com_mod.tnNo;
+  int nFacesLS = com_mod.nFacesLS;
+
+  Ag_.resize(tDof, tnNo);
+  Yg_.resize(tDof, tnNo);
+  Dg_.resize(tDof, tnNo);
+  res_.resize(nFacesLS);
+  incL_.resize(nFacesLS);
+}
+
+//------------------------
+// step
+//------------------------
+/// @brief Execute one Newton iteration loop for the current time step
+bool Integrator::step() {
+  using namespace consts;
+
+  auto& com_mod = simulation_->com_mod;
+  auto& cm_mod = simulation_->cm_mod;
+  auto& cep_mod = simulation_->get_cep_mod();
+
+  auto& An = com_mod.An;
+  auto& Yn = com_mod.Yn;
+  auto& Dn = com_mod.Dn;
+
+  int& cTS = com_mod.cTS;
+  int& cEq = com_mod.cEq;
+
+  #define n_debug_integrator_step
+  #ifdef debug_integrator_step
+  DebugMsg dmsg(__func__, com_mod.cm.idcm());
+  dmsg.banner();
+  #endif
+
+  // Inner loop for Newton iteration
+  inner_count_ = 1;
+  int reply;
+  int iEqOld;
+
+  // Looping over Newton iterations
+  while (true) {
+    #ifdef debug_integrator_step
+    dmsg << "---------- Inner Loop " + std::to_string(inner_count_) << " -----------" << std::endl;
+    dmsg << "cEq: " << cEq;
+    dmsg << "com_mod.eq[cEq].sym: " << com_mod.eq[cEq].sym;
+    #endif
+
+    auto istr = "_" + std::to_string(cTS) + "_" + std::to_string(inner_count_);
```

Comment:
`istr` could be a class variable that's updated in every time step and Newton iteration